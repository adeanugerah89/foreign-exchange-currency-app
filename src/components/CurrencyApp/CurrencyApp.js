import React, { Component } from 'react';
import axios from 'axios';
import CurrencyListApp from '../CurrencyListApp/CurrencyListApp';
import { Button, FormControl } from 'react-bootstrap';
import '../../styles/CurrencyAppCss.css';
import '../../styles/CurrencyListAppCss.css';

class CurrencyApp extends Component {
    constructor(props) {
        super(props);
        this.state = {
            addCurrencies: '',
            dataList: [
                {
                    baseValue: "14901.50",
                    id: 1538372038451,
                    key: "IDR",
                    value: "14901.50",
                },
                {
                    baseValue: "0.86",
                    id: 1538372238662,
                    key: "EUR",
                    value: 0.86
                },
                {
                    baseValue: "0.77",
                    id: 1538372337662,
                    key: "GBP",
                    value: 0.77
                },
                {
                    baseValue: "1.37",
                    id: 1538372397202,
                    key: "SGD",
                    value: 1.37
                }
            ],
            dataSource: []
        };

        this.handleAddCurrencies = this.handleAddCurrencies.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.deleteItem = this.deleteItem.bind(this);
    }

    componentDidMount() {
        axios.get(`https://api.exchangeratesapi.io/latest?base=USD`)
          .then((response) => {
              let dataRates = response.data
              for(let value in dataRates.rates){
                  dataRates.rates[value] = dataRates.rates[value].toFixed(2)
                }

              this.setState({
                dataSource: dataRates
              })
          })
          .catch((error) => console.log(error))
     }

    componentDidUpdate (prevProps){
        if(this.props !== prevProps){
            let newStateValue = this.state.dataList.map(data => data.baseValue * this.props.inputValue)
            let newDataList = []
            this.state.dataList.forEach((data, idx) => {
                 newDataList.push({
                     key: data.key,
                     value:  newStateValue[idx],
                     baseValue:data.baseValue,
                     id: data.id
                 })
            })

            this.setState({
                dataList: newDataList
            })
        }

        return true
    }

    deleteItem(id) {
        let filteredItems = this.state.dataList.filter((item) => {
            return (item.id !== id)
        });

        this.setState({
            dataList: filteredItems
        })
    }

    handleAddCurrencies(event) {
        this.setState({addCurrencies: event.target.value});
    }

    handleSubmit(event) {
        let dataSource = this.state.dataSource.rates
        let addCurrencies = this.state.addCurrencies.toUpperCase()
        let existingKey = this.state.dataList.filter(data => data.key === addCurrencies)
        if (event.target.value !== "" && existingKey.length === 0){
            let newAddDataListItem = {}
            for (let key in dataSource) {
                if (key === addCurrencies) {
                    let value = this.props.inputValue ? this.props.inputValue * dataSource[key] : dataSource[key]
                    newAddDataListItem = {
                        key: key,
                        value: value,
                        baseValue: dataSource[key],
                        id: Date.now()
                    };
                }
            }

            if(Object.keys(newAddDataListItem).length !== 0) {
                this.setState((prev) => {
                    return { dataList: prev.dataList.concat(newAddDataListItem) }
                })
            }else {
                alert('Data not found, please check your input text')
            }

        }
        event.preventDefault();
        this.setState({addCurrencies: ''});
     }

    render(){
        return (
            <div className="currencyListMain">
                <CurrencyListApp
                    entries={this.state.dataList}
                    delete={this.deleteItem}
                />
                <div className="header">
                    <form onSubmit={this.handleSubmit}>
                        <FormControl
                          type="text"
                          value={this.state.addCurrencies}
                          placeholder="add more currencies"
                          onChange={this.handleAddCurrencies}
                        />
                        <Button bsStyle="primary" type="submit">Submit</Button>
                    </form>
                </div>
            </div>
        )
    }
}

export default CurrencyApp;
