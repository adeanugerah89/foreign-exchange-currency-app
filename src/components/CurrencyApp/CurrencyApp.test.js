import React from 'react';
import {mount, render, shallow} from 'enzyme';
import '../setupTest';
import { spy } from 'sinon';

import CurrencyApp from './CurrencyApp';

describe('<CurrencyApp /> test', () => {

    it('renders correctly', () => {
        const wrapper = shallow(<CurrencyApp />);
        expect(wrapper).toMatchSnapshot();
    });

    it('<CurrencyApp /> component should have state', () => {
        const wrapper = shallow(<CurrencyApp />);
        expect(wrapper.state).toHaveLength(1);
    });

    it('<CurrencyApp /> should have property entries', () => {
        const wrapper = shallow(<CurrencyApp />);
        expect(wrapper.props().children[0].props).toHaveProperty('entries');
    });

    it('<CurrencyApp /> component pass props', () => {
        const wrapper = shallow(<CurrencyApp />);
        expect(wrapper.props().children[0].props.entries.length).toBeGreaterThan(1)
    });

    it('<CurrencyApp /> component props contain object', () => {
        const wrapper = shallow(<CurrencyApp />);
        expect(typeof(wrapper.props().children[0].props.entries)).toEqual('object')
    });

});
