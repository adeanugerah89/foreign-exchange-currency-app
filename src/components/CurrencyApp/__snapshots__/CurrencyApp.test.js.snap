// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyApp /> test renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CurrencyApp />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <CurrencyListApp
          delete={[Function]}
          entries={
                    Array [
                              Object {
                                "baseValue": "14901.50",
                                "id": 1538372038451,
                                "key": "IDR",
                                "value": "14901.50",
                              },
                              Object {
                                "baseValue": "0.86",
                                "id": 1538372238662,
                                "key": "EUR",
                                "value": 0.86,
                              },
                              Object {
                                "baseValue": "0.77",
                                "id": 1538372337662,
                                "key": "GBP",
                                "value": 0.77,
                              },
                              Object {
                                "baseValue": "1.37",
                                "id": 1538372397202,
                                "key": "SGD",
                                "value": 1.37,
                              },
                            ]
          }
/>,
        <div
          className="header"
>
          <form
                    onSubmit={[Function]}
          >
                    <FormControl
                              bsClass="form-control"
                              componentClass="input"
                              onChange={[Function]}
                              placeholder="add more currencies"
                              type="text"
                              value=""
                    />
                    <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="primary"
                              disabled={false}
                              type="submit"
                    >
                              Submit
                    </Button>
          </form>
</div>,
      ],
      "className": "currencyListMain",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "delete": [Function],
          "entries": Array [
            Object {
              "baseValue": "14901.50",
              "id": 1538372038451,
              "key": "IDR",
              "value": "14901.50",
            },
            Object {
              "baseValue": "0.86",
              "id": 1538372238662,
              "key": "EUR",
              "value": 0.86,
            },
            Object {
              "baseValue": "0.77",
              "id": 1538372337662,
              "key": "GBP",
              "value": 0.77,
            },
            Object {
              "baseValue": "1.37",
              "id": 1538372397202,
              "key": "SGD",
              "value": 1.37,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <form
            onSubmit={[Function]}
>
            <FormControl
                        bsClass="form-control"
                        componentClass="input"
                        onChange={[Function]}
                        placeholder="add more currencies"
                        type="text"
                        value=""
            />
            <Button
                        active={false}
                        block={false}
                        bsClass="btn"
                        bsStyle="primary"
                        disabled={false}
                        type="submit"
            >
                        Submit
            </Button>
</form>,
          "className": "header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FormControl
                bsClass="form-control"
                componentClass="input"
                onChange={[Function]}
                placeholder="add more currencies"
                type="text"
                value=""
/>,
              <Button
                active={false}
                block={false}
                bsClass="btn"
                bsStyle="primary"
                disabled={false}
                type="submit"
>
                Submit
</Button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "form-control",
                "componentClass": "input",
                "onChange": [Function],
                "placeholder": "add more currencies",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "active": false,
                "block": false,
                "bsClass": "btn",
                "bsStyle": "primary",
                "children": "Submit",
                "disabled": false,
                "type": "submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": [Function],
            },
          ],
          "type": "form",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <CurrencyListApp
            delete={[Function]}
            entries={
                        Array [
                                    Object {
                                      "baseValue": "14901.50",
                                      "id": 1538372038451,
                                      "key": "IDR",
                                      "value": "14901.50",
                                    },
                                    Object {
                                      "baseValue": "0.86",
                                      "id": 1538372238662,
                                      "key": "EUR",
                                      "value": 0.86,
                                    },
                                    Object {
                                      "baseValue": "0.77",
                                      "id": 1538372337662,
                                      "key": "GBP",
                                      "value": 0.77,
                                    },
                                    Object {
                                      "baseValue": "1.37",
                                      "id": 1538372397202,
                                      "key": "SGD",
                                      "value": 1.37,
                                    },
                                  ]
            }
/>,
          <div
            className="header"
>
            <form
                        onSubmit={[Function]}
            >
                        <FormControl
                                    bsClass="form-control"
                                    componentClass="input"
                                    onChange={[Function]}
                                    placeholder="add more currencies"
                                    type="text"
                                    value=""
                        />
                        <Button
                                    active={false}
                                    block={false}
                                    bsClass="btn"
                                    bsStyle="primary"
                                    disabled={false}
                                    type="submit"
                        >
                                    Submit
                        </Button>
            </form>
</div>,
        ],
        "className": "currencyListMain",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "delete": [Function],
            "entries": Array [
              Object {
                "baseValue": "14901.50",
                "id": 1538372038451,
                "key": "IDR",
                "value": "14901.50",
              },
              Object {
                "baseValue": "0.86",
                "id": 1538372238662,
                "key": "EUR",
                "value": 0.86,
              },
              Object {
                "baseValue": "0.77",
                "id": 1538372337662,
                "key": "GBP",
                "value": 0.77,
              },
              Object {
                "baseValue": "1.37",
                "id": 1538372397202,
                "key": "SGD",
                "value": 1.37,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <form
              onSubmit={[Function]}
>
              <FormControl
                            bsClass="form-control"
                            componentClass="input"
                            onChange={[Function]}
                            placeholder="add more currencies"
                            type="text"
                            value=""
              />
              <Button
                            active={false}
                            block={false}
                            bsClass="btn"
                            bsStyle="primary"
                            disabled={false}
                            type="submit"
              >
                            Submit
              </Button>
</form>,
            "className": "header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  onChange={[Function]}
                  placeholder="add more currencies"
                  type="text"
                  value=""
/>,
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="primary"
                  disabled={false}
                  type="submit"
>
                  Submit
</Button>,
              ],
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "form-control",
                  "componentClass": "input",
                  "onChange": [Function],
                  "placeholder": "add more currencies",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "active": false,
                  "block": false,
                  "bsClass": "btn",
                  "bsStyle": "primary",
                  "children": "Submit",
                  "disabled": false,
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Submit",
                "type": [Function],
              },
            ],
            "type": "form",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
